src/nvim/drawline.c

```diff
@@ -275,7 +278,50 @@ static void draw_virt_text(win_T *wp, buf_T *buf, int col_off, int *end_col, int
     if (decor_virt_pos(item) && item->draw_col == -1) {
       bool updated = true;
       VirtTextPos pos = decor_virt_pos_kind(item);
-      if (pos == kVPosRightAlign) {
+
+      if (do_eol && pos == kVPosEndOfLineRightAlign) {
+        int eolOffset = 0;
+        if (totalWidthOfEolRightAlignedVirtText == 0) {
+          // Look ahead to the remaining decor items
+          for (int j = i; j < end; j++) {
+            /// A future decor to be handled in this function's call
+            DecorRange *lookaheadItem = &slots[indices[j]].range;
+
+            if (lookaheadItem->start_row != state->row) {
+              continue;
+            }
+
+            if (!decor_virt_pos(lookaheadItem)) {
+              continue;
+            }
+
+            if (lookaheadItem->draw_col != -1) {
+              continue;
+            }
```

Comment:
this seems a bit verbose, and isn't really typical for this codebase.
```suggestion
            if (lookaheadItem->start_row != state->row
                || !decor_virt_pos(lookaheadItem)
                || lookaheadItem->draw_col != -1) {
              continue;
            }
```